### PR TITLE
Prefer using ConfigProperties over Config in AgentListener

### LINKED ISSUE #4925: Improve configuration story of AgentListener API
Relates to #4831 and #4791 and [SDK #3924](https://github.com/open-telemetry/opentelemetry-java/issues/3924)

So presently, the `AgentListener` has methods like:
```
  default void beforeAgent(Config config, AutoConfiguredOpenTelemetrySdk autoConfiguredOpenTelemetrySdk) {}
```

The config instance passed to this method _seems_ redundant because the `AutoConfiguredOpenTelemetrySdk` has a `getConfig()` method on it. However, the type on the sdk is `ConfigMap` and not `Config`. We like the `Config` object (superior to ConfigMap because of the defaulting methods) and need it in our distribution.

This issue is then to see if we can migrate the `AutoConfiguredOpenTelemetrySdk` to `Config` or otherwise introduce a wrapper or similar, in order to eliminate the extra config param from the `AgentListener` methods.

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Resolves #4925 by removing the `Config` parameter from `AgentListener` -- now that `ConfigProperties` has get-or-default methods we can use them in the listener implementations.

~~And I think we can resolve 6022 too~~. We can't really get rid of `Config`, as we must have a global configuration singleton that's accessible from application/instrumentation code (can't use `ConfigProperties` because it's not part of the API; and thus not located in the bootstrap CL); it also needs to work without the agent, in a library instrumentation scenario.

I think it makes sense to use `ConfigProperties` everywhere in the agent SPIs, so that they're consistent with the autoconfigure SPIs. In the instrumentation code, we can keep using the global `Config`.

### CODE REVIEW

**Review by trask — APPROVED** [review]
nice!

File: javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentListener.java
Comment (mateuszrzeszutek) [reply javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentListener.java:45]: > are you thinking to drop the default after the next release?

Yes, that was my intention 👍 

> and do you think we're confident after that to mark it with `@FunctionInterface` or better not to?

Hm, it's an SPI, it can't be a lambda function -- probably better not to add it.

File: instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java
Comment (mateuszrzeszutek) [inline instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java:48]: I found one more mismatch between `ConfigProperties` and `Config` -- in javaagent, we use the dash character in property names fairly often; turns out that SDK `ConfigProperties` does not support this kind of thing.
@jack-berg @anuraaga do you think we should bridge that gap in the SDK?

Comment (jack-berg) [reply instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java:48]: I'm not familiar with the history of why `DefaultConfigProperties` replaces `-`, `_` with `.`. On the surface level it seems reasonable to perform the same normalization on keys when retrieving config values. Would defer to @jkwatson and @anuraaga though.

Comment (anuraaga) [reply instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java:48]: SDK had just one at the time (and now two I think) properties with hyphens so it was just the easiest implementation given `ConfigProperties` was only generally used from autoconfigure implementation itself. Since then, it's exposed in many APIs and mysterious normalization is probably a bug :P

File: javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentListener.java
Comment (trask) [reply javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentListener.java:45]: > it's an SPI, it can't be a lambda function

🤦‍♂️ 👍

File: instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java
Comment (mateuszrzeszutek) [reply instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java:48]: > I'm not familiar with the history of why `DefaultConfigProperties` replaces `-`, `_` with `.`.

The `_` is replaced as a part of env var -> system prop name conversion. `-` was probably copied from the agent code (we replaced it with `.` to bridge the gap between the env vars and system properties).

> mysterious normalization is probably a bug :P

Is it? True, hyphens are not used in the SDK, but they're commonly used in the javaagent property names (I did some archeology work and found this comment explaining the reasoning behind it: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1747#issuecomment-733154848). Do we want `ConfigProperties` to behave similar to javaagent's `Config` with regard to this? I think that if we want to use `ConfigProperties` in javaagent SPIs to (for similarity with SDK SPIs) we should make `ConfigProperties` work with the kebab-case properties.

Comment (trask) [reply instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java:48]: > I did some archeology work

😄

> Do we want `ConfigProperties` to behave similar to javaagent's `Config` with regard to this?

I think so (pending approval from @open-telemetry/java-maintainers)

Comment (anuraaga) [reply instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java:48]: > Is it? True, hyphens are not used in the SDK, but they're commonly used in the javaagent property names

Right I mean it's a bug in the SDK that it is normalizing away hyphens. Back when the autoconfigure's config was only used within autoconfigure, it wasn't a big deal that if a user specifies `my.custom-property` on the command line, it is only accessible as `.get("my.custom.property")`, but since it's exposed to users a lot now, it seems buggy. I can't remember what exactly was easier to go with the hacky normalization but we shouldn't do it.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java
Comment (anuraaga) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java:30]: Is it possible to make this class just delegate to `ConfigProperties`? I realized we have some confusing SPI because we can customize `ConfigProperties` in SDK and `Config` here. If we need to expose a factory to construct `DefaultConfigProperties` for the case where there's no javaagent or SDK, we could probably figure something out, but I guess we do need to converte these SPIs somehow.

Comment (anuraaga) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java:30]: Also I guess this is the last chance to look into possibly moving `ConfigProperties` to opentelemetry-api (existing one could be deprecated and just delegate to it). If we need config from instrumentation, having two, one for SDK and one for instrumentation, seems like it is making us struggle a lot.

Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java:30]: > possibly moving `ConfigProperties` to opentelemetry-api

That would be perfect, I think.

Comment (jack-berg) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java:30]: > possibly moving ConfigProperties to opentelemetry-api

The idea being that instrumentation would only need to rely on `opentelemetry-api` instead of the additional `autoconfigure-spi` dependency?

Comment (jkwatson) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java:30]: I'm not a fan of putting this interface into the OTel API. This is 100% an SDK construct, used only by the SDK. It shouldn't be added to the API.

Comment (anuraaga) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java:30]: We currently have basically the exact same concept used in instrumentation as well as `Config` here, so currently have that duplication which is particularly confusing when customizing the config. Library instrumentation could be purely programmatic - I think we decided to fallback to system a while back but could reconsider this. But if we think that OTel instrumentation "is supposed to" support config through environment, then it stops being an SDK-only construct and we may want to think through more options on it.

@mateuszrzeszutek In the meantime, as with many of our classes, maybe we should just move `Config` to `internal` for now 😅

Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java:30]: Notes from today's special topics meeting:

1. We'll add hyphen normalization to SDK's `ConfigProperties`
2. `ConfigProperties` will be extracted to its own, separate `opentelemetry-config` module in the SDK repo. It'll be used both by agent and the SDK autoconfigure. The global `Config` instance will be moved from `instrumentation-api` to the `javaagent-extension-api` bootstrap part (it still needs to be available in the agent), `Instrumenter`s and telemetry builders will have a new method that allows passing a `ConfigProperties` instance.
3. In case 1 & 2 are not finished before the release, we'll just move `Config` to internal so that it does not block marking the `instrumentation-api` release as RC.

File: instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java
Comment (mateuszrzeszutek) [reply instrumentation/runtime-metrics/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/RuntimeMetricsInstaller.java:48]: I've added an issue for that in the SDK repo: https://github.com/open-telemetry/opentelemetry-java/issues/4562
For now, I think we can merge this PR as it is and just leave a TODO note that this will be fixed separately.

**Comment by mateuszrzeszutek** [comment]
I think this is ready to be merged; I'll continue working on making the agent SPIs use `ConfigProperties` and deprecating `Config` in the following PRs.

**Pull request merged**
